src/custom/state/orders/updater.ts

```diff
@@ -0,0 +1,176 @@
+}
+
+export function EventUpdater(): null {
+  const { chainId, library } = useActiveWeb3React()
+  // console.log('EventUpdater::library', library)
+  // console.log('EventUpdater::chainId', chainId)
+
+  const lastBlockNumber = useBlockNumber()
+  const lastCheckedBlock = useLastCheckedBlock({ chainId })
+  console.log('EventUpdater::lastCheckedBlock', lastCheckedBlock)
+  console.log('EventUpdater::lastBlockNumber', lastBlockNumber)
+
+  const dispatch = useDispatch<AppDispatch>()
+
+  // show popup on confirm
+  // for displaying fulfilled orders
+  const addPopup = useAddPopup()
+
+  const getLogsRetry = useMemo(() => {
+    if (!library) return null
+    return constructGetLogsRetry(library)
+  }, [library])
+
+  useEffect(() => {
+    if (!chainId || !library || !getLogsRetry || !lastBlockNumber) return
+    ;(window as any).getLogsRetry = (fromBlock: number, toBlock: number) => {
+      // to play around with in console
+      return getLogsRetry({
+        fromBlock,
+        toBlock,
+        topics: TransferEventTopics
+      })
+    }
+
+    // don't check for fromBlock > toBlock
+    if (lastCheckedBlock + 1 > lastBlockNumber) return
+
+    const getPastEvents = async () => {
+      console.log('EventUpdater::getLogs', {
+        fromBlock: lastCheckedBlock + 1,
+        toBlock: lastBlockNumber,
+        // address: '0x',
+        topics: TransferEventTopics
+      })
+
+      const logs = await getLogsRetry({
+        fromBlock: lastCheckedBlock + 1,
+        toBlock: lastBlockNumber,
+        // address: '0x',
+        topics: TransferEventTopics
+      })
+
+      // console.log('logs', logs)
+      logs.forEach(log => {
+        try {
+          // const { from, to, amount } = decodeTransferEvent(log)
+          // const { from, to, amount, id } = decodeTransferEvent(log)
+
+          console.log('EventUpdater::Detected transfer of token in block', log.blockNumber)
+          // console.log('EventUpdater::Detected transfer of token since last time', log.address, { from, to, amount })
+          // dispatch(fulfillOrder({ chainId, id, fulfillmentTime: from log,if: from log }))
+        } catch (e) {}
+      })
+
+      // TODO: extend addPopup to accept whatever we want to show for Orders
+      if (logs.length > 0) {
+        const firstBlock = logs[0].blockNumber
+        const lastBlock = logs[logs.length - 1].blockNumber
+
+        const blocksRangeStr = firstBlock === lastBlock ? `block ${firstBlock}` : `blocks ${firstBlock} - ${lastBlock}`
+        // Sample popup for events
+        addPopup(
+          {
+            txn: {
+              hash: logs[0].transactionHash,
+              success: true,
+              summary: `EventUpdater::Detected ${logs.length} token Transfers in ${blocksRangeStr}`
+            }
+          },
+          logs[0].transactionHash
+        )
+      }
+
+      // SET lastCheckedBlock = lastBlockNumber
+      dispatch(updateLastCheckedBlock({ chainId, lastCheckedBlock: lastBlockNumber }))
```

Comment:
I wonder if this should be a unisque action, that at the same time:
- Finalize several orders
- Update the block

I mean. Doesn't it makes sense to update all the state with all our info. You are doing a Promise.all so is not that you are consuming on the go. So u are triggering re-draws in the middle. Also because it feels unnatural to me an action called `updateLastCheckedBlock`. I think the block (and date) should be part of the payload of a `finalizeOrders` actions

WDYT?

Reply:
I was planning to use `batch` from `react-redux`. Takes care of rerenders

Btw, multiple dispatches happen exactly like that in `transactions/updater`

Reply:
ok, but wouldn't it be better to just model it differently? less actions, and less missuse. You receive the full payload that would help u transit to the new state, instead of depending on not breaking before `updateLastCheckedBlock` and then getting duplicates

Reply:
I don't really see the less misuse angle.But I will add _one action to rule them all_ together with fetching block timestamps for `fulfillmentTime`
In a separate PR

Reply:
My point on the misuse is, you can call `updateLastCheckedBlock` indepently of the actual finalization of orders. In my head there was one action happening: `fulfillOrders({  orderIds,  lastCheckedBlock }`

Anyways, is not that imp

Reply:
But we **do** call `updateLastCheckedBlock` independently
```ts
logs.forEach(log => {
  dispatch(fulfillOrder())
})

dispatch(updateLastCheckedBlock())
```
Guess what happens when `logs` is empty

Reply:
u call it independently, but functionally at the same time. If this would be a relational database, I would do it in a single transaction, rolling back the whole thing

I wouldn't like to end up with some fulfilled order, and some not, and maybe the block unupdated. Probably it wouldn't matter since I guess fullfill is idempotent, but I guess u get the point

Why I said this, is because you await for all the events anyways, so you already have all the events resolved. Anyways, was just a comment, u can leave it like this